flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java

```diff
@@ -208,7 +215,13 @@ public void recycle(MemorySegment segment) {
         }
 
         try {
-            inputChannel.notifyBufferAvailable(numAddedBuffers);
+            if (releasedFloatingBuffer == null) {
+                if (!inputChannel.isReleased()) {
+                    inputChannel.notifyBufferAvailable(1);
+                }
```

Comment:
Is this `if` check a relevant change to this bug fix? If not, could you please extract it to a separate hotfix commit in this PR to avoid confusions?

Reply:
In the original code, the parameter of notifyBufferAvailable() may be 0 or 1. When it is 0, there is no need to call notifyBufferAvailable(). When it is 1, no floating buffer is released, and !inputChannel.isReleased() , The if judgment here is just to keep it in line with the original logic.

Reply:
But I don't see what `!inputChannel.isReleased()` has to do with this? Previously code was not checking for that, and always calling either `notifyBufferAvailable(0)` or `notifyBufferAvailable(1)`. I get it, that in the current form, calling `notifyBufferAvailable(0)` doesn't make sense, but if there is no problem with skipping the `if (!inputChannel.isReleased())` check, I would just skip it for the sake of simplicity.

On the other hand, if there is a problem, I presume the problem must have been on the master before, and it's unrelated to this bug fix?

Reply:
The explanation is a bit awkward, and I also didn't add this judgment at first.
In the original code `(numAddedBuffers == 0)` == `(!inputChannel.isReleased() && getAvailableBufferSize()> numRequiredBuffers)`, if `inputChannel.isReleased()`, the exclusive buffer will be added directly to the `globalPool` instead of adding to exclusive buffer queue. Therefore, even if `releasedFloatingBuffer` is null, we don't necessarily should call `notifyBufferAvailable(1)`.
Previously, this logic was recorded through `numAddedBuffers`, but I removed `numAddedBuffers` in this modification, since we don't need `numAddedBuffers` to record the return value of `addExclusiveBuffer()` any more.
Do you think it would be better to keep the original `numAddedBuffers`?

Reply:
Ok, thanks for the explanation. Now I have finally understood the problem 😅 

0. I'm not sure if we can we drop this second `isReleased()` check and always call `notifyBufferAvailable()`? It would simplify the code, but I don't know if that's correct thing to do.
1. In your current proposal, I would be worried that `inputChannel.isReleased()` was `false`, but maybe it has changed to `true` during the execution of this method. Maybe just check `isReleased()` only once and store it's result on a `boolean isReleased` local variable and re-use it?
2a. Maybe we can restructure the code? Either to something like that:
```
public void recycle(MemorySegment segment) {
  if (inputChannel.isReleased()) {
    recycleWithReleasedChannel(segment); //new private method
  }
  else {
    recycleWithChannel(segment); //new private method
  }
```
2b. or just add a return statement:
```
                if (inputChannel.isReleased()) {
                    globalPool.recycleMemorySegments(Collections.singletonList(segment));
                    return;
                } else {
```
2b is not as readable as 2a, but 2a. would be a bit more verbosed.

What do you think?

Reply:
Thank you for your detailed reply. ☺️
0. I have tested that `notifyBufferAvailable()` is always called, which will cause the test case `CreditBasedPartitionRequestClientHandlerTest::testReadBufferResponseBeforeReleasingChannel()` to fail.
1. This does have the problem of atomicity.
2. If we are not deliberately following the original logic, I think 2b has no readability problem. The special case `inputChannel.isReleased()` is processed and returned at the beginning of the function, which is more natural and easier to understand. 2b is also an inline implementation of 2a, but the code is split into two parts by lock. Fortunately, the special case processing is short enough, so the overall logic is relatively smooth.
I will add a commit to implement 2b.